src/rules/noUnusedVariableRule.ts

```diff
@@ -33,18 +39,58 @@ class NoUnusedVariablesWalker extends Lint.RuleWalker {
     private skipParameterDeclaration: boolean;
     private skipVariableDeclaration: boolean;
 
+    // after a this.walk is completed, this.reactImport will be set to a NamespaceImport
+    // iff:
+    //
+    // + a react option has been provided to the rule and
+    // + an import of a module that matches one of OPTION_REACT_MODULES is found, to a
+    //   NamespaceImport named OPTION_REACT_NAMESPACE_IMPORT_NAME
+    //
+    // e.g.
+    //
+    // import * as React from "react/addons";
+    //
+    // If reactImport is defined when a walk is completed, we need to have:
+    //
+    // a) seen another usage of React and/or
+    // b) seen a JSX identifier
+    //
+    // otherwise a usage error will be reported
+    private reactImport: ts.NamespaceImport;
+    private reactUsed: boolean;
+    private seenJsx: boolean;
+
     constructor(sourceFile: ts.SourceFile, options: Lint.IOptions, languageService: ts.LanguageService) {
         super(sourceFile, options);
         this.languageService = languageService;
         this.skipVariableDeclaration = false;
         this.skipParameterDeclaration = false;
+        this.reactImport = undefined;
+        this.reactUsed = false;
+        this.seenJsx = false;
+    }
+
+    public visitSourceFile(node: ts.SourceFile): void {
```

Comment:
Happy to go with whatever you prefer here. 

However, in our project we enforce the `"call-signature"` option of the `typedef` rule. Reason being, it's clearer if function signatures are explicit:

```typescript
function A() {
    return 5;
}
```